src/lib/updateContractInfo.ts

```diff
@@ -0,0 +1,191 @@
+ *
+ * @param {tokenAddress} tokenAddress - The address of the token.
+ * @param {blockNumber} blockWithTransactions - The block (with transactions) when updating contractinfo by block.
+
+ * @returns {Promise<void>} This function does not return any useful value.
+ */
+export async function updateContractInfoByTokenAddress(
+  {
+    tokenAddress,
+    transferBlockNumber,
+    goal,
+    fullUpdate = false
+  }: IContractInfoJobDetailsByTokenAddress,
+  blockWithTransactions?: BlockWithTransactions
+): Promise<void> {
+  const address = tokenAddress.toLowerCase()
+  const startTime = Date.now()
+  stats.increment('update_contract_called')
+
+  try {
+    const [contractInfo, created] = await ContractInfo.findOrCreate({
+      where: {
+        address
+      }
+    })
+
+    if (!created && goal === ProcessingGoal.BACKFILL) {
+      stats.increment('update_token_backfill')
+      return
+    }
+
+    const updateSpec =
+      fullUpdate ||
+      created ||
+      !contractInfo.updatedMetaInfoAt ||
+      differenceInDays(Date.now(), contractInfo.updatedMetaInfoAt) >=
+        CONTRACT_INFO_MAX_AGE_IN_DAYS
+
+    const updateBlockMetrics =
+      !!transferBlockNumber &&
+      (!contractInfo.lastTransactionBlock ||
+        transferBlockNumber > contractInfo.lastTransactionBlock)
+
+    if (!updateSpec && !updateBlockMetrics && contractInfo.ethBalance) {
+      stats.increment('update_token_not_stale')
+      return
+    }
+
+    const client = await getEthClient()
+    const contract = new Contract(address, abi, client)
+
+    const [contractName, symbol, decimals, tokenType, ethBalance, block] =
+      await Promise.all([
+        updateSpec ? safeCall<string>(contract, 'name') : contractInfo.name,
+        updateSpec ? safeCall<string>(contract, 'symbol') : contractInfo.symbol,
+        updateSpec ? safeCall<BN>(contract, 'decimals') : contractInfo.decimals,
+        updateSpec ? getContractSpec(address) : contractInfo.tokenType,
+        client.getBalance(address),
+        updateBlockMetrics
+          ? blockWithTransactions || client.getBlock(transferBlockNumber)
+          : undefined
+      ])
+
+    if (updateSpec) {
+      contractInfo.name = contractName || knownContracts[address]?.name
+      contractInfo.symbol = symbol || knownContracts[address]?.symbol
+      contractInfo.tokenType = tokenType
+      contractInfo.decimals = decimals
+        ? +decimals.toString()
+        : knownContracts[address]?.decimals
+
+      contractInfo.updatedMetaInfoAt = new Date()
+    }
+
+    if (updateBlockMetrics) {
+      contractInfo.lastTransactionBlock = transferBlockNumber
+      if (block?.timestamp) {
+        const blockTimestanp = new Date(block.timestamp * 1000)
+        if (
+          !contractInfo.lastTransactionAt ||
+          isAfter(blockTimestanp, contractInfo.lastTransactionAt)
+        ) {
+          contractInfo.lastTransactionAt = blockTimestanp
+        }
+      }
+    }
```

Comment:
Just to make sure I understand. The reasoning behind having `updateSpec` and `updateBlockMetrics` is that there's no reason to keep updating the spec every time there's a transaction with a contract because that kind of info is not likely to have changed, whereas we definitely want to update the blockMetrics because we want to have the latest block time stamp?

Reply:
correct. meta change should maybe we checked once a day. updateBlockMetrics should run only if block is higher than one we have on record now.

Reply:
cool cool